src/pressure/VCycleMG.cpp

```diff
@@ -7,120 +7,102 @@
 
 
-// =============================== Constructor ===============================
-// *****************************************************************************
-/// \brief  Constructor
-/// \param  out     pressure
-/// \param  b       rhs
-// *****************************************************************************
-VCycleMG::VCycleMG(Field const &out, Field const &b) {
+VCycleMG::VCycleMG(Field const &out, Field const &b) :
+        m_levels(Domain::getInstance()->get_levels()),
+        m_n_cycle(Parameters::getInstance()->get_int("solver/pressure/n_cycle")),
+        m_n_relax(Parameters::getInstance()->get_int("solver/pressure/diffusion/n_relax")),
+        m_dt(Parameters::getInstance()->get_real("physical_parameters/dt")),
+        m_w(Parameters::getInstance()->get_real("solver/pressure/diffusion/w")) {
 #ifndef BENCHMARKING
     m_logger = Utility::create_logger(typeid(this).name());
 #endif
+    Parameters *params = Parameters::getInstance();
+
+    std::string diffusion_type = params->get("solver/pressure/diffusion/type");
+    if (diffusion_type == DiffusionMethods::Jacobi) {
+        m_diffusion_max_iter = static_cast<size_t>(params->get_int("solver/pressure/diffusion/max_solve"));
+        m_smooth_function = &VCycleMG::call_smooth_jacobi;
+        m_solve_function = &VCycleMG::call_solve_jacobi;
+    } else if (diffusion_type == DiffusionMethods::ColoredGaussSeidel) {
+        m_diffusion_max_iter = static_cast<size_t>(params->get_int("solver/pressure/diffusion/max_iter"));
+        m_smooth_function = &VCycleMG::call_smooth_colored_gauss_seidel;
+        m_solve_function = &VCycleMG::call_solve_colored_gauss_seidel;
+    } else {
+#ifndef BENCHMARKING
+        m_logger->critical("Diffusion method not yet implemented! Simulation stopped!");
+#endif
+        // TODO(issue 6) Error handling
+    }
+    m_diffusion_tol_res = params->get_real("solver/pressure/diffusion/tol_res");
 
-    auto params = Parameters::getInstance();
-    auto domain = Domain::getInstance();
-
-    m_levels = domain->get_levels();
-    m_cycles = params->get_int("solver/pressure/n_cycle");
-    m_relaxs = params->get_int("solver/pressure/diffusion/n_relax");
-
-    m_dsign = -1.;
-    m_w = 2. / 3.;
-    m_w = params->get_real("solver/pressure/diffusion/w");
-
-    Field *out_err1 = new Field(out.get_type(), 0.0, 0, domain->get_size());
-    Field *out_tmp = new Field(out.get_type(), 0.0, 0, domain->get_size());
-    Field *b_res1 = new Field(b.get_type(), 0.0, 0, domain->get_size());
-
-    out_err1->copy_data(out);
-    out_tmp->copy_data(out);
-    b_res1->copy_data(b);
+    m_residuum0 = new Field *[m_levels];
+    m_residuum1 = new Field *[m_levels + 1];
+    m_error0 = new Field *[m_levels];
+    m_error1 = new Field *[m_levels + 1];
+    m_mg_temporal_solution = new Field *[m_levels + 1];
 
+    // copies of out and b to prevent aliasing
     // residuum
+    auto *b_res1 = new Field(b);
     b_res1->copyin();
```

Comment:
same here, because the copy constructor also copies to the GPU.
(see ll. 60,65,77,82,87,92)